File: crown-bin/src/args/hash.rs
use std::fmt::Display;

use clap::{Parser, ValueEnum};

#[derive(Debug, Parser)]
pub struct ArgsHash {
    #[clap(default_value = "sha256", required = true)]
    pub algorithm: HashAlgorithm,
    #[clap(short, long, default_value_t = true)]
    pub text: bool,
    #[clap(short, long, default_value_t = false)]
    pub binary: bool,
    #[clap(short, long, default_value_t = false)]
    pub check: bool,
    #[clap(required = true)]
    pub files: Vec<String>,
    #[clap(long, default_value_t = false)]
    pub hmac: bool,
    #[clap(long, default_value = None)]
    pub key: Option<String>,
    /// output hash size(in bytes) for XOF.
    #[clap(long, default_value_t = 32)]
    pub length: usize,
}

#[derive(Debug, Parser)]
pub struct Md5 {
    #[clap(short, long, default_value_t = true)]
    pub text: bool,
    #[clap(short, long, default_value_t = false)]
    pub binary: bool,
    #[clap(short, long, default_value_t = false)]
    pub check: bool,
    pub files: Vec<String>,
}

#[derive(Debug, Default, Clone, Copy, ValueEnum, strum_macros::EnumString)]
#[clap(rename_all = "kebab-case")]
pub enum HashAlgorithm {
    Md2,
    Md4,
    #[default]
    Md5,
    Sha1,
    Sha224,
    Sha256,
    Sha384,
    Sha512,
    #[clap(name = "sha512-224")]
    Sha512224,
    #[clap(name = "sha512-256")]
    Sha512256,
    #[clap(name = "sha3-224")]
    Sha3224,
    #[clap(name = "sha3-256")]
    Sha3256,
    #[clap(name = "sha3-384")]
    Sha3384,
    #[clap(name = "sha3-512")]
    Sha3512,
    Blake2b256,
    Blake2b384,
    Blake2b512,
    Blake2s128,
    Blake2s256,
    Blake2b,
    Blake2s,
    Shake128,
    Shake256,
    #[cfg(feature = "cuda")]
    Md5Cuda,
    #[cfg(feature = "cuda")]
    Sha256Cuda,
}

impl HashAlgorithm {
    #[cfg(feature = "cuda")]
    pub fn is_cuda(&self) -> bool {
        match self {
            #[cfg(feature = "cuda")]
            Self::Md5Cuda | Self::Sha256Cuda => true,
            _ => false,
        }
    }
}

impl Display for HashAlgorithm {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let v = match self {
            Self::Md2 => "md2",
            Self::Md4 => "md4",
            Self::Md5 => "md5",
            Self::Sha1 => "sha1",
            Self::Sha224 => "sha224",
            Self::Sha256 => "sha256",
            Self::Sha384 => "sha384",
            Self::Sha512 => "sha512",
            Self::Sha512224 => "sha512-224",
            Self::Sha512256 => "sha512-256",
            Self::Sha3224 => "sha3-224",
            Self::Sha3256 => "sha3-256",
            Self::Sha3384 => "sha3-384",
            Self::Sha3512 => "sha3-512",
            Self::Blake2b256 => "blake2b-256",
            Self::Blake2b384 => "blake2b-384",
            Self::Blake2b512 => "blake2b-512",
            Self::Blake2s128 => "blake2s128",
            Self::Blake2s256 => "blake2s256",
            Self::Blake2b => "blake2b",
            Self::Blake2s => "blake2s",
            Self::Shake128 => "shake128",
            Self::Shake256 => "shake256",
            #[cfg(feature = "cuda")]
            Self::Md5Cuda => "md5-cuda",
            #[cfg(feature = "cuda")]
            Self::Sha256Cuda => "sha256-cuda",
        };

        write!(f, "{}", v)
    }
}

File: crown-playground/lib/hash.ts
import { Hash } from 'crown-wasm';

export type HashAlgorithm =
  | 'md2'
  | 'md4'
  | 'md5'
  | 'sha1'
  | 'sha224'
  | 'sha256'
  | 'sha384'
  | 'sha512'
  | 'sha512_224'
  | 'sha512_256'
  | 'sha3_224'
  | 'sha3_256'
  | 'sha3_384'
  | 'sha3_512'
  | 'shake128'
  | 'shake256'
  | 'blake2s'
  | 'blake2b'
  | 'sm3';

export function createHash(
  algorithm: HashAlgorithm,
  hmacKey?: Uint8Array,
): Hash {
  const isHmac = hmacKey !== undefined;

  if (algorithm === 'blake2s' || algorithm === 'blake2b') {
    return Hash[`new_${algorithm}`](hmacKey, 32);
  }

  const methodName =
    `new_${algorithm}${isHmac ? '_hmac' : ''}` as keyof typeof Hash;
  const method = Hash[methodName] as any;

  if (typeof method !== 'function') {
    throw new Error(
      `Unsupported algorithm: ${algorithm}${isHmac ? ' with HMAC' : ''}`,
    );
  }

  return isHmac ? method(hmacKey) : method();
}

export function supportsHmac(algorithm: HashAlgorithm): boolean {
  if (algorithm === 'blake2s' || algorithm === 'blake2b') {
    return true;
  }

  const hmacMethodName = `new_${algorithm}_hmac` as keyof typeof Hash;
  return typeof Hash[hmacMethodName] === 'function';
}

export function getAvailableAlgorithms(): {
  value: HashAlgorithm;
  label: string;
  supportsHmac: boolean;
}[] {
  const algorithms: {
    value: HashAlgorithm;
    label: string;
    supportsHmac: boolean;
  }[] = [
    // MD family
    { value: 'md2', label: 'MD2', supportsHmac: supportsHmac('md2') },
    { value: 'md4', label: 'MD4', supportsHmac: supportsHmac('md4') },
    { value: 'md5', label: 'MD5', supportsHmac: supportsHmac('md5') },

    // SHA-1
    { value: 'sha1', label: 'SHA-1', supportsHmac: supportsHmac('sha1') },

    // SHA-2 family
    { value: 'sha224', label: 'SHA-224', supportsHmac: supportsHmac('sha224') },
    { value: 'sha256', label: 'SHA-256', supportsHmac: supportsHmac('sha256') },
    { value: 'sha384', label: 'SHA-384', supportsHmac: supportsHmac('sha384') },
    { value: 'sha512', label: 'SHA-512', supportsHmac: supportsHmac('sha512') },
    {
      value: 'sha512_224',
      label: 'SHA-512/224',
      supportsHmac: supportsHmac('sha512_224'),
    },
    {
      value: 'sha512_256',
      label: 'SHA-512/256',
      supportsHmac: supportsHmac('sha512_256'),
    },

    // SHA-3 family
    {
      value: 'sha3_224',
      label: 'SHA3-224',
      supportsHmac: supportsHmac('sha3_224'),
    },
    {
      value: 'sha3_256',
      label: 'SHA3-256',
      supportsHmac: supportsHmac('sha3_256'),
    },
    {
      value: 'sha3_384',
      label: 'SHA3-384',
      supportsHmac: supportsHmac('sha3_384'),
    },
    {
      value: 'sha3_512',
      label: 'SHA3-512',
      supportsHmac: supportsHmac('sha3_512'),
    },

    // SHAKE (extendable-output functions)
    {
      value: 'shake128',
      label: 'SHAKE128',
      supportsHmac: supportsHmac('shake128'),
    },
    {
      value: 'shake256',
      label: 'SHAKE256',
      supportsHmac: supportsHmac('shake256'),
    },

    // BLAKE2 family
    {
      value: 'blake2s',
      label: 'BLAKE2s',
      supportsHmac: supportsHmac('blake2s'),
    },
    {
      value: 'blake2b',
      label: 'BLAKE2b',
      supportsHmac: supportsHmac('blake2b'),
    },

    // SM3 (Chinese national standard)
    { value: 'sm3', label: 'SM3', supportsHmac: supportsHmac('sm3') },
  ];

  return algorithms;
}
